test/ARCHITECTURE.md

```diff
@@ -0,0 +1,267 @@
+usually based on a distribution's cloud image customized with:
+
+* A well-known password for the admin/root user
+* Test SSH keys for access
+* Test packages required to test Cockpit
+* A build chroot with Cockpit's build dependencies to build the to be tested
+  Cockpit source inside the virtual machine offline. This allows a developer on
+  Fedora to easily prepare a Debian test image without having to install Debian
+  build tools.
+* Disabling system services which interfere with testing
+
+To replicate a user, Cockpit is tested in a browser controlled using the
+[Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
+(CDP) which is supported by Firefox and Chromium based browsers.
+
+The Python test framework in `test/common` is responsible for setting up the
+test environment, running tests and reporting of the test output.
+
+Diagram of the interaction of Browser/Machine/CDP/Test Framework.
+```mermaid
+graph TD;
+    id[Test Framework] <-->|CDP| Browser;
+    id[Test Framework] <-->|SSH| id1[Virtual Machine];
+```
+
+## Integration Tests
+
+Cockpit's tests can be run via three different entry points:
+
+* `test/verify/check-$page` - run a single or multiple unit test(s)
+* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties)
+* `test/run` - run tests in Continuous Integration (CI)
+
+We will start with how a single integration test is run and then explore the test
+scheduler and CI setup.
+
+### Test runtime
+
+The base of a Cockpit integration test looks as following:
+
+```python
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.machine.execute("rm /usr/share/cockpit/apps")
+        self.browser.login_and_go("/apps")
+        self.browser.wait_not_visible("#apps")
+
+if __name__ == '__main__':
+    testlib.test_main()
+```
+
+In Cockpit there are two types of tests, `destructive` and `nondestructive`
+tests. Destructive tests do something to test virtual machine which makes it
+unable to run another test afterwards or requires another virtual machine for
+testing. The test above is a `destructive` test which is the default, a non
+`destructive` test makes sure any destructive action is restored after the test
+has run as can be seen below. `nondestructive` tests where introduced to speed
+up testing, as rebooting and shutting down a machine for every tests incurs a
+significant penalty of ~ 10-30 seconds per test.
+
+```python
+@testlib.nondestructive
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.restore_dir("/usr/share/cockpit/apps")
+        self.machine.execute("rm /usr/share/cockpit/apps")
+        self.browser.login_and_go("/apps")
+        self.browser.wait_not_visible("#apps")
+
+if __name__ == '__main__':
+    testlib.test_main()
+```
+
+The test above would be invoked via `./test/verify/check-apps TestApps.testBasic`
+and would execute as can be seen in the diagram below:
+```mermaid
+sequenceDiagram
```

Comment:
Please checkout the rendered diagram, this comes out all wrong and garbled :cry: